src/npg_irods/utilities.py

```diff
@@ -44,6 +50,170 @@ def _print(path, writer):
         print(path, file=writer)
 
 
+def check_checksums(
+    reader, writer, num_threads=1, num_clients=1, print_pass=True, print_fail=False
+) -> bool:
+    """Read iRODS data object paths from a file and check that each one has correct
+    checksums and checksum metadata, printing the results to a writer.
+
+    The conditions for checksums and checksum metadata of a data object to be correct
+    are:
+
+    - The data object must have a checksum set for each valid replica.
+    - The checksums for all replicas must have the same value.
+    - The data object must have one, and only one, checksum AVU in its metadata.
+    - The checksum AVU must have the same value as the replica checksums.
+
+    Args:
+        reader: A file supplying iRODS data object paths to check, one per line.
+        writer: A file where checked paths will be written, one per line.
+        num_threads: The number of Python threads to use. Defaults to 1.
+        num_clients: The number of baton clients to use, Defaults to 1.
+        print_pass: Print the paths of objects passing the check. Defaults to True.
+        print_fail: Print the paths of objects failing the check. Defaults to False.
+
+    Returns:
+        True if all checks are done.
+    """
+    with client_pool(num_clients) as bp:
+
+        def fn(i: int, path: str):
+            success = False
+
+            p = path.strip()
+            try:
+                obj = DataObject(p, pool=bp)
+                if has_matching_checksum_metadata(obj):
+                    log.info("Checksums correct", item=i, path=obj)
+                    if print_pass:
+                        _print(p, writer)
+                else:
+                    checksums = [
+                        avu.value
+                        for avu in obj.metadata()
+                        if avu.attribute == DataFile.MD5.value
+                    ]
+                    checksums.sort()
+                    log.warn(
+                        "Checksum metadata do not match",
+                        item=i,
+                        path=obj,
+                        checksum=obj.checksum(),
+                        metadata=checksums,
+                    )
+
+                    if print_fail:
+                        _print(p, writer)
+
+                success = True
+
+            except RodsError as re:
+                log.error(re.message, code=re.code)  # Much more informative diagnostic
```

Comment:
```suggestion
                log.error(re.message, item=i, code=re.code)  # Much more informative diagnostic
```
Should this be the same as the other occurences of this error message?

Reply:
Good catch, thanks.